docs/proposals/lvmd-within-topolvm-node.md

```diff
@@ -0,0 +1,312 @@
+	fs.StringVar(&cfgFilePath, "config", filepath.Join("/etc", "topolvm", "lvmd.yaml"), "config file")
+	fs.BoolVar(&command.Containerized, "container", false, "Run within a container")
+    //...
+
+	fs.AddGoFlagSet(goflags)
+}
+```
+
+After this, we can use the flags taken from the lvmd binary and start it within the command:
+
+```go
+package cmd
+
+import (
+	"context"
+	"errors"
+	"net"
+	"os"
+	"time"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"github.com/cybozu-go/log"
+	"github.com/spf13/viper"
+	"github.com/topolvm/topolvm"
+	topolvmlegacyv1 "github.com/topolvm/topolvm/api/legacy/v1"
+	topolvmv1 "github.com/topolvm/topolvm/api/v1"
+	clientwrapper "github.com/topolvm/topolvm/client"
+	"github.com/topolvm/topolvm/controllers"
+	"github.com/topolvm/topolvm/driver"
+	"github.com/topolvm/topolvm/lvmd"
+	"github.com/topolvm/topolvm/lvmd/proto"
+	"github.com/topolvm/topolvm/runners"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+	"sigs.k8s.io/yaml"
+	//+kubebuilder:scaffold:imports
+)
+
+var (
+	scheme   = runtime.NewScheme()
+	setupLog = ctrl.Log.WithName("setup")
+)
+
+func init() {
+	utilruntime.Must(clientgoscheme.AddToScheme(scheme))
+
+	utilruntime.Must(topolvmv1.AddToScheme(scheme))
+	utilruntime.Must(topolvmlegacyv1.AddToScheme(scheme))
+	//+kubebuilder:scaffold:scheme
+}
+
+func subMain() error {
+	nodename := viper.GetString("nodename")
+	if len(nodename) == 0 {
+		return errors.New("node name is not given")
+	}
+
+	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&config.zapOpts)))
+
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+		Scheme:             scheme,
+		MetricsBindAddress: config.metricsAddr,
+		LeaderElection:     false,
+	})
+	if err != nil {
+		setupLog.Error(err, "unable to start manager")
+		return err
+	}
+	client := clientwrapper.NewWrappedClient(mgr.GetClient())
+	apiReader := clientwrapper.NewWrappedReader(mgr.GetAPIReader(), mgr.GetClient().Scheme())
+
+	var lvclnt proto.LVServiceClient
+	var vgclnt proto.VGServiceClient
+
+	if config.llvmd {
+		if err := loadConfFile(cfgFilePath); err != nil {
+			return err
+		}
+		dcm := lvmd.NewDeviceClassManager(config.lvmd.DeviceClasses)
+		ocm := lvmd.NewLvcreateOptionClassManager(config.lvmd.LvcreateOptionClasses)
+		_, lvclnt, _, vgclnt = lvmd.NewLocal(dcm, ocm)
+	} else {
+		dialer := &net.Dialer{}
+		dialFunc := func(ctx context.Context, a string) (net.Conn, error) {
+			return dialer.DialContext(ctx, "unix", a)
+		}
+		conn, err := grpc.Dial(config.lvmdSocket, grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithContextDialer(dialFunc))
```

Comment:
`conn` is not used in the following lines.

In the latest code, it is passed to `NewLogicalVolumeReconciler`.
https://github.com/topolvm/topolvm/blob/52dbdcce42eba0509a4a0e94eaa04e9571f74cd9/pkg/topolvm-node/cmd/run.go#L76

You should change how to initialize `lvcontroller` depending on the value of `config.llvmd`.